internal/xds/translator/translator.go

```diff
@@ -113,29 +114,57 @@ func Translate(ir *ir.Xds) (*types.ResourceVersionTable, error) {
 		tCtx.AddXdsResource(resource.ClusterType, xdsCluster)
 
 		// Search for an existing listener, if it does not exist, create one.
-		xdsListener := findXdsListener(tCtx, tcpListener.Address, tcpListener.Port)
+		xdsListener := findXdsListener(tCtx, tcpListener.Address, tcpListener.Port, core.SocketAddress_TCP)
 		if xdsListener == nil {
-			xdsListener = buildXdsListener(tcpListener.Name, tcpListener.Address, tcpListener.Port)
+			xdsListener = buildXdsTCPListener(tcpListener.Name, tcpListener.Address, tcpListener.Port)
 			tCtx.AddXdsResource(resource.ListenerType, xdsListener)
 		}
-
+		// todo: huabing zhao we should check the conflicted listeners: multiple TCP listeners or TCP and HTTP listeners
+		// should not listen on the same port
 		if err := addXdsTCPFilterChain(xdsListener, tcpListener, xdsCluster.Name); err != nil {
 			return nil, err
 		}
 	}
+
+	for _, udpListener := range ir.UDP {
+		// 1:1 between IR UDPListener and xDS Cluster
```

Comment:
Is the 1:1 b/c UDPRoute does not support any of the [UDP matcher rules](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api#extension-category-envoy-matching-network-input), e.g. source IP?

Reply:
@danehans Envoy can route UDP diagrams by their soure IP, but the G: ateway API doesn't support that.

Envoy UDP route: https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/udp_filters/udp_proxy#routing

Gateway UDPRoute API: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.UDPRouteSpec